File: android/app/src/main/java/biz/am2/swiftbible/data/DevotionalHistory.kt
package biz.am2.swiftbible.data

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject

/**
 * Compact rolling log of devotional opens so the All Voices and Series
 * Completionist hidden badges can evaluate against recent history. Mirrors
 * iOS DevotionalHistory. Backed by a dedicated SharedPreferences key so
 * it stays separate from UserPreferences and won't bloat that snapshot.
 */
object DevotionalHistory {
    private const val PREFS_NAME = "devotional_history_v1"
    private const val KEY_HISTORY = "history"
    private const val MAX_ENTRIES = 200

    data class View(
        val timestamp: Long,
        val track: String?,
        val seriesName: String?,
        val seriesPart: Int?,
    )

    fun record(context: Context, track: String?, seriesName: String?, seriesPart: Int?) {
        val prefs = prefs(context)
        val history = load(prefs).toMutableList()
        history.add(View(System.currentTimeMillis(), track, seriesName, seriesPart))
        while (history.size > MAX_ENTRIES) history.removeAt(0)
        save(prefs, history)
    }

    /** Distinct AI track names viewed in the trailing window. */
    fun tracksInLast(context: Context, days: Int): Set<String> {
        val cutoff = System.currentTimeMillis() - days.toLong() * 24L * 60 * 60 * 1000
        return load(prefs(context))
            .filter { it.timestamp >= cutoff }
            .mapNotNull { it.track }
            .toSet()
    }

    /** Series-name → set of parts viewed; used for Series Completionist. */
    fun seriesProgress(context: Context): Map<String, Set<Int>> {
        val map = mutableMapOf<String, MutableSet<Int>>()
        for (view in load(prefs(context))) {
            val name = view.seriesName ?: continue
            val part = view.seriesPart ?: continue
            map.getOrPut(name) { mutableSetOf() }.add(part)
        }
        return map
    }

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private fun load(prefs: SharedPreferences): List<View> {
        val raw = prefs.getString(KEY_HISTORY, null) ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                View(
                    timestamp = o.getLong("ts"),
                    track = o.optString("track").takeIf { it.isNotEmpty() },
                    seriesName = o.optString("series").takeIf { it.isNotEmpty() },
                    seriesPart = if (o.has("part")) o.getInt("part") else null,
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun save(prefs: SharedPreferences, history: List<View>) {
        val arr = JSONArray()
        for (v in history) {
            val o = JSONObject().apply {
                put("ts", v.timestamp)
                v.track?.let { put("track", it) }
                v.seriesName?.let { put("series", it) }
                v.seriesPart?.let { put("part", it) }
            }
            arr.put(o)
        }
        prefs.edit().putString(KEY_HISTORY, arr.toString()).apply()
    }
}

File: ios/swiftbible/Views/Progress/DevotionalHistory.swift
//
//  DevotionalHistory.swift
//  swiftbible
//

import Foundation

/// Compact rolling log of devotional opens so the All Voices and Series
/// Completionist badges can evaluate against recent history without
/// reaching into PostHog. Records carry just enough metadata to support
/// the two queries; everything else is dropped.
struct DevotionalView: Codable, Equatable {
    let date: Date
    let track: String?
    let seriesName: String?
    let seriesPart: Int?
}

@MainActor
enum DevotionalHistory {
    private static let storageKey = "devotionalHistoryV1"
    private static let maxEntries = 200

    static func record(track: String?, seriesName: String?, seriesPart: Int?) {
        var history = load()
        history.append(DevotionalView(date: Date(), track: track, seriesName: seriesName, seriesPart: seriesPart))
        // Keep history bounded — enough for any "last N weeks" query, never
        // unbounded on heavy users.
        if history.count > maxEntries {
            history.removeFirst(history.count - maxEntries)
        }
        save(history)
    }

    static func load() -> [DevotionalView] {
        guard let data = UserDefaults.standard.data(forKey: storageKey),
              let decoded = try? JSONDecoder().decode([DevotionalView].self, from: data) else {
            return []
        }
        return decoded
    }

    /// Distinct AI track names viewed in the trailing window.
    static func tracksInLast(days: Int) -> Set<String> {
        guard let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: Date()) else { return [] }
        return Set(load()
            .filter { $0.date >= cutoff }
            .compactMap { $0.track })
    }

    /// Series-name → set of parts the user has viewed. Used for the
    /// Series Completionist badge (parts 1..4 within one named series).
    static func seriesProgress() -> [String: Set<Int>] {
        var progress: [String: Set<Int>] = [:]
        for view in load() {
            guard let name = view.seriesName, let part = view.seriesPart else { continue }
            progress[name, default: []].insert(part)
        }
        return progress
    }

    private static func save(_ history: [DevotionalView]) {
        guard let encoded = try? JSONEncoder().encode(history) else { return }
        UserDefaults.standard.set(encoded, forKey: storageKey)
    }
}
